File: leetcode/python/15.3Sum.py
class Test:

  @staticmethod
  def threeSum(nums):
    result = []
    nums.sort()
    for i in range(len(nums)):
      if i == 0 or nums[i] != nums[i-1]:
        start = i+1
        end = len(nums) - 1
        while start < end :
          s = nums[i] + nums[start] + nums[end]
          if s == 0:
            temp = [nums[i],nums[start],nums[end]]
            result.append(temp)
            startVal = nums[start]
            endVal = nums[end]
            while start < end and startVal == nums[start]:
              start+=1
            while end > start and endVal == nums[end]:
              end-=1
          elif s < 0:
            start+=1
          else:
            end-=1
    return result  
            
print(Test.threeSum([-1, 0, 1, 2, -1, -4]))

File: leetcode/js/15.3Sum.js
class Test{

    static threeSum(nums){
      var result=[]
      nums.sort((a,b)=>a-b)
      for(let i=0;i<nums.length;i++){
        if(i==0||nums[i]!=nums[i-1]){
          let start = i+1
          let end = nums.length - 1
          while(start<end){
            let s = nums[i] + nums[start] + nums[end]
            if(s==0){
                let temp = [nums[i],nums[start],nums[end]]
                result.push(temp)
                let startVal = nums[start]
                let endVal = nums[end]
                while(start < end && startVal == nums[start]){
                   start+=1        
                }
                while(end > start && endVal == nums[end]){
                  end-=1
                }
            }else if(s<0){start+=1}
            else{end-=1}
          }
        }
      }
      return result
    }
    
    }
    
    console.log(Test.threeSum([-1, 0, 1, 2, -1, -4]))